### PR TITLE
Add PEM decoding support

### PR DESCRIPTION
**Pull request by LeviSchuck**
This adds a new top level function `decode_pem`, and a subsequent function `as_key` on the result type for use with `sign` and `verify`.

This contributes to #76 and #77 
However it does not include X.509 certificate support.

I have an idea on how to do that but this should handle many cases that users want as PEM files are the common output from `openssl`

It supports both PKCS#1 and PKCS#8 RSA keys, it supports PKCS#8 EC keys, however it doesn't assert that the curve ID but within JWT contexts the curve is predefined by the spec.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -17,3 +17,5 @@ serde = "1.0"
 ring = { version = "0.16.5", features = ["std"] }
 base64 = "0.10"
 chrono = "0.4"
+pem = "0.6"
```
Comment (Keats) [inline Cargo.toml]: hmm this add `failure` to the list of sub-dependencies, I'll check if the author would accept a PR removing it.

File: README.md [hunk 2]
```diff
@@ -113,8 +113,10 @@ This library currently supports the following:
 - ES384
 
 ### RSA
-`jsonwebtoken` can only read DER encoded keys currently. If you have openssl installed,
-you can run the following commands to obtain the DER keys from PKCS#1 (ie with `BEGIN RSA PUBLIC KEY`) .pem.
+`jsonwebtoken` can read DER and PEM encoded keys.
+
+If you want to use DER encoded keys review the following.
```
Comment (Keats) [inline README.md]: this sentence seems like it's missing stuff?

File: Cargo.toml (hunk 1 above)
Comment (LeviSchuck) [reply Cargo.toml]: > How did you generate the PEM keys? I will add all the defaut ones from jwt.io later on to ensure everything is fine.

All of the key data is derived from what was already in the repo. I renamed the `private_rsa_key.pem` to its corresponding `_pkcs8` type. I've gone through my zsh history for exact answers.

The PKCS#1 keys are the base64 encoded `.der` keys already in the repository with the corresponding PEM tag `PRIVATE RSA KEY` or `PUBLIC RSA KEY`.

The PKCS#8 RSA public key was created from the private key with an openssl command.
`openssl pkcs8 -topk8 -in private_rsa_key.pem -out private_rsa_key_pkcs8.pem -nocrypt`

The PKCS#8 EC private key was created from the `.pk8` file already in the repository.
`base64 public_ecdsa_key.pk8 |  awk '{gsub(/.{64}/,"&\n")}1' > public_ecdsa_key.pem` followed by manually adding the PEM Tagging `-----BEGIN PRIVATE KEY-----` and `-----END PRIVATE KEY-----` around the existing file.

And the public key was then derived from that
`openssl ec -in private_ecdsa_key.pem -pubout -out public_ecdsa_key.pem`

**Comment by LeviSchuck** [comment]
> How did you generate the PEM keys? I will add all the defaut ones from jwt.io later on to ensure everything is fine.

All of the key data is derived from what was already in the repo. I renamed the `private_rsa_key.pem` to its corresponding `_pkcs8` type. I've gone through my zsh history for exact answers.

The PKCS#1 keys are the base64 encoded `.der` keys already in the repository with the corresponding PEM tag `PRIVATE RSA KEY` or `PUBLIC RSA KEY`.

The PKCS#8 RSA public key was created from the private key with an openssl command.
`openssl pkcs8 -topk8 -in private_rsa_key.pem -out private_rsa_key_pkcs8.pem -nocrypt`

The PKCS#8 EC private key was created from the `.pk8` file already in the repository.
`base64 public_ecdsa_key.pk8 |  awk '{gsub(/.{64}/,"&\n")}1' > public_ecdsa_key.pem` followed by manually adding the PEM Tagging `-----BEGIN PRIVATE KEY-----` and `-----END PRIVATE KEY-----` around the existing file.

And the public key was then derived from that
`openssl ec -in private_ecdsa_key.pem -pubout -out public_ecdsa_key.pem`

File: README.md (hunk 2 above)
Comment (LeviSchuck) [reply README.md]: I've updated the readme a little.

**Comment by LeviSchuck** [comment]
> I will probably change the API to make it easy for users but all the effort it very appreciated.

Sure, I'm not set on how the API came out, though I struggled with lifetimes. There's the challenges I experienced that resulted in this API.

1. PEM Decoding allocates its own `Vec<u8>` for the base64 contents.
2. ASN1 Parsing references the slice made from the `Vec<u8>` above
3. The `Key::(Der|Pkcs8)` enums use a slice reference, which may either be directly from the PEM decoded contents or from a subsection as selected from the ASN1 parsing. 

This is the first real rust and opensource contribution I've made so I appreciate the critique you give.

**Comment by LeviSchuck** [comment]
Your PR seems to have landed Keats, and is now on crate
https://github.com/jcreekmore/pem-rs/pull/19
https://crates.io/crates/pem

I'll update the PR by bumping the pem version

File: Cargo.toml (hunk 1 above)
Comment (LeviSchuck) [reply Cargo.toml]: It's now 0.7 on crates and this is now updated.

**Comment by Keats** [comment]
Thanks!

> This is the first real rust and opensource contribution I've made so I appreciate the critique you give.

I'll give a better review then but that will have to wait the weekend/next week.

File: src/pem_decoder.rs [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+  PKCS1,
+  PKCS8,
+}
+
+#[derive(Debug)]
+#[derive(PartialEq)]
+enum Classification {
+  EC,
+  RSA,
+}
```
Comment (Keats) [inline src/pem_decoder.rs]: I would put all of those at the top above the struct.

File: src/pem_decoder.rs [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+
+#[derive(Debug)]
+#[derive(PartialEq)]
+enum PemEncodedWith {
+  PKCS1,
+  PKCS8,
+}
+
+#[derive(Debug)]
+#[derive(PartialEq)]
```
Comment (Keats) [inline src/pem_decoder.rs]: You can (and should) put those derive in a single command: `#[derive(Debug, PartialEq)]`

File: src/pem_decoder.rs [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+      }
+      simple_asn1::ASN1Block::ObjectIdentifier(_, oid) => {
+        if oid == ec_public_key_oid {
+          return Option::Some(Classification::EC);
+        } else if oid == rsa_public_key_oid {
+          return Option::Some(Classification::RSA);
+        }
+      }
+      _ => {
+        println!("Ignoring {:?}", asn1_entry);
```
Comment (Keats) [inline src/pem_decoder.rs]: you can remove that

File: src/pem_decoder.rs [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+  for asn1_entry in asn1.iter() {
+    match asn1_entry {
+      simple_asn1::ASN1Block::Sequence(_, entries) => {
+        if let Some(classification) = classify_pem(entries) {
+          return Some(classification);
+        }
+      }
+      simple_asn1::ASN1Block::ObjectIdentifier(_, oid) => {
+        if oid == ec_public_key_oid {
+          return Option::Some(Classification::EC);
```
Comment (Keats) [inline src/pem_decoder.rs]: No need for `Option::`

File: src/pem_decoder.rs [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+        } else if oid == rsa_public_key_oid {
+          return Option::Some(Classification::RSA);
+        }
+      }
+      _ => {
+        println!("Ignoring {:?}", asn1_entry);
+      }
+    }
+  }
+  return Option::default();
```
Comment (Keats) [inline src/pem_decoder.rs]: you can replace that line by `None`

File: src/pem_decoder.rs (hunk 5 above)
Comment (LeviSchuck) [reply src/pem_decoder.rs]: oh yes definitely, that was a debugging helper.

**Comment by Keats** [comment]
Note: I'll merge it in the pem-handling branch first to experiment with the API and to make it easy for you to follow all the changes/give feedback.

**Comment by LeviSchuck** [comment]
I look forward to it.

**Pull request merged**